File: tree/100. 相同的树.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {boolean}
 */
 var isSameTree = function(p, q) {
    // 深度优先遍历
    // 如果p,q 节点都为null，返回true
    if(p === null && q === null){
        return true;
    }
    // 如果p,q节点只有一个为null返回false;
    else if(p === null || q === null){
        return false;
    }
    // 如果p,q两节点值不相等直接返回false;
    else if(p.val !== q.val){
        return false;
    }
    // 如果p,q节点值相等，递归，当左子树递归返回值和右子树递归返回值都为true时返回true;
    else{
        return isSameTree(p.left,q.left) && isSameTree(p.right,q.right);
    }

};

File: Python/100. 相同的树.py
# -*- codeing = utf-8 -*-
# @Time : 2023-8-19 11:45
# @Author : 刘永奇
# @File : 100. 相同的树.py
# @Software : PyCharm
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        # p,q都为空返回True
        if not p and not q:return True
        # p,q有一个不为空则返回False
        elif not p or not q: return False
        # 都不为空比较值如果不相等返回False
        elif p.val != q.val: return False
        # 相等则深度优先遍历左右子树
        else: return self.isSameTree(p.left,q.left) and self.isSameTree(p.right,q.right)